core/sail/memory/src/main/java/org/eclipse/rdf4j/sail/memory/model/MemStatementIterator.java

```diff
@@ -8,17 +8,17 @@
 package org.eclipse.rdf4j.sail.memory.model;
 
 import java.util.Arrays;
+import java.util.NoSuchElementException;
 
 import org.eclipse.rdf4j.common.iteration.CloseableIteration;
-import org.eclipse.rdf4j.common.iteration.LookAheadIteration;
 import org.eclipse.rdf4j.sail.SailException;
 
 /**
  * A StatementIterator that can iterate over a list of Statement objects. This iterator compares Resource and Literal
  * objects using the '==' operator, which is possible thanks to the extensive sharing of these objects in the
  * MemoryStore.
  */
-public class MemStatementIterator<X extends Exception> extends LookAheadIteration<MemStatement, X> {
+public class MemStatementIterator implements CloseableIteration<MemStatement, SailException> {
```

Comment:
If this is the case, do you see any legit use cases for `LookAheadIteration` as an (abstract) class, still? If not, should we mark it deprecated?

Reply:
I think that we should sacrifice DRY for performance when it comes to hot code such as the iteration code. Itable stubbing might be less of an issue when we move away from having both the Iteration and CloseableIteration interfaces. Stubbing also only happens on the first method invocation of an object, so long lived iterations will not really be affected much.

If possible I would prefer trying to create some static methods in a helper class for common use cases such as a LookAheadIteration, instead of using inheritance. I also want to play with keeping inheritance but removing polymorphic methods, essentially marking all iteration classes as final and adding abstract methods for the cases where we need to override a method.